src/librustc_mir/interpret/machine.rs

```diff
@@ -254,6 +254,15 @@ pub trait Machine<'mir, 'tcx>: Sized {
         kind: Option<MemoryKind<Self::MemoryKind>>,
     ) -> (Cow<'b, Allocation<Self::PointerTag, Self::AllocExtra>>, Self::PointerTag);
 
+    /// Called to notify the machine before a deallocation occurs
+    fn before_deallocation(
+        _memory_extra: &Self::MemoryExtra,
```

Comment:
Is not not possible to provide `&mut` here?

Reply:
I can, should I? I don't think I need it to be mutable for miri diagnostics.

Reply:
Yes please do. I think it makes sense that a hook corresponding to a mutating operation may mutate extra state.